src/ucp/wireup/wireup_cm.c

```diff
@@ -41,6 +41,57 @@ int ucp_ep_init_flags_has_cm(unsigned ep_init_flags)
                                UCP_EP_INIT_CM_WIREUP_SERVER));
 }
 
+static const char* ucp_cm_used_idx_to_cm_name(ucp_context_h context,
+                                              ucp_rsc_index_t cms_used_idx,
+                                              char *buf, size_t max)
+{
+    ucp_rsc_index_t cmpt_index = context->config.cm_cmpt_idxs[cms_used_idx];
+
+    ucs_snprintf_safe(buf, max, "%s", context->tl_cmpts[cmpt_index].attr.name);
+    return buf;
+}
+
+/*
+ * The main thread progress part of attempting connecting the client to the server
+ * through the next available cm.
+ */
+static unsigned ucp_cm_client_try_next_cm_progress(void *arg)
+{
+    ucp_ep_h ucp_ep       = arg;
+    ucp_worker_h worker   = ucp_ep->worker;
+    ucp_context_h context = worker->context;
+    ucp_wireup_ep_t *cm_wireup_ep;
+    ucs_status_t status;
+    uct_ep_h uct_cm_ep;
+    char cm_name_prev[UCT_TL_NAME_MAX], cm_name_next[UCT_TL_NAME_MAX];
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    cm_wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    ucs_assert_always(cm_wireup_ep != NULL);
+    uct_cm_ep = ucp_ep_get_cm_uct_ep(ucp_ep);
+    ucs_assert_always(uct_cm_ep != NULL);
+
+    /* try connecting through another cm (next one in the priority list) */
+    cm_wireup_ep->cms_used_idx++;
+    ucp_wireup_ep_unset_next_ep(cm_wireup_ep);
+    uct_ep_destroy(uct_cm_ep);
```

Comment:
pls use `ucp_wireup_ep_extract_next_ep` or move destroy inside `ucp_wireup_ep_unset_next_ep`, so `uct_cm_ep` related code above becomes redundant.

Reply:
why not keep the destroy of `uct_cm_ep` outside the `ucp_wireup_ep_unset_next_ep` function?

same as `ucp_wireup_ep_set_next_ep` doesn't create the `next_ep`

Reply:
because this function almost duplicates existing one